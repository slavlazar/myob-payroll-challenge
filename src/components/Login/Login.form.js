import React from 'react';
import PropTypes from 'prop-types';
import { Form } from 'formik';
import * as Yup from 'yup';

import { Input, Button } from '../shared/ui';

/**
 * Login schema for this form
 * Contains validation for fields username and password
 */
export const LoginSchema = Yup.object().shape({
  username: Yup.string().required('Please enter your username'),
  password: Yup.string().required('Please enter your password'),
});

/**
 * Login form generated by Formik for our Login screen
 * @param {object} Params from Formik for this Login form
 */
const LoginForm = ({
  values,
  touched,
  errors,
  isSubmitting,
  handleChange,
  handleBlur,
  handleSubmit,
  isValid,
  loginError,
}) => (
  <Form>
    <Input
      id="username"
      type="text"
      name="username"
      label="Username"
      onBlur={handleBlur}
      onChange={handleChange}
      value={values.username}
      errorMessage={
        touched.username && errors.username ? errors.username : null
      }
    />
    <Input
      id="password"
      type="password"
      name="password"
      label="Password"
      onBlur={handleBlur}
      onChange={handleChange}
      value={values.password}
      errorMessage={
        touched.password && errors.password ? errors.password : null
      }
    />
    <Button
      id="login-form-button"
      colour="primary"
      type="submit"
      aria-label="Sign in"
      block
      onClick={handleSubmit}
      disabled={!isValid || (isSubmitting && !loginError)}
    >
      Sign in
    </Button>
  </Form>
);

LoginForm.propTypes = {
  /** Values from Formik */
  values: PropTypes.object.isRequired,
  /** Object of touched elements from Formik */
  touched: PropTypes.object.isRequired,
  /** Function to handle form submission */
  handleSubmit: PropTypes.func.isRequired,
  /** Function to handle field change */
  handleChange: PropTypes.func.isRequired,
  /** Function to handle field blur */
  handleBlur: PropTypes.func.isRequired,
  /** Boolean to determine if form passes validation schema */
  isValid: PropTypes.bool.isRequired,
  isSubmitting: PropTypes.bool,
  errors: PropTypes.shape({
    username: PropTypes.string,
    password: PropTypes.string,
  }),
  loginError: PropTypes.string,
};

LoginForm.defaultProps = {
  isSubmitting: false,
  errors: {},
  loginError: null,
};

export default LoginForm;
