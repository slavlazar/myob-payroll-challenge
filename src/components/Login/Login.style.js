import styled, { keyframes } from 'styled-components';

import { media } from '../../utils/style-utils';

const animated = keyframes`
  from {
    opacity: 0;
    transform: scale3d(0.3, 0.3, 0.3);
  }

  50% {
    opacity: 1;
  }
`;

export const LogoContainer = styled.div`
  width: 10rem;
`;

export const CardContainer = styled.div`
  width: 35rem;
  animation-name: ${animated};
  animation-fill-mode: both;
  animation-duration: 600ms;

  ${media.tablet`width: 60rem`};
  ${media.desktop`width: 70rem`};
`;
