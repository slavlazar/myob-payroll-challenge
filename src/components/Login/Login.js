import React from 'react';
import PropTypes from 'prop-types';
import { Formik } from 'formik';

import { LogoContainer, CardContainer } from './Login.style';
import { Logo, Card, Heading, Alert, FullScreenFlex } from '../shared/ui';
import LoginForm, { LoginSchema } from './Login.form';
import { generateInitialValuesFromSchema } from '../../utils/helpers';

/**
 * Login component displaying our view with login form generated by Formik
 * @param {object} Login component params from our Auth state
 */
const Login = ({ handleLogin, loading, error }) => (
  <FullScreenFlex>
    <LogoContainer>
      <Logo />
    </LogoContainer>
    <Heading textAlign="center">Sign in to Pay Slip Generator</Heading>
    <CardContainer>
      <Card loading={loading}>
        {error && (
          <Alert
            className="login-alert"
            type="danger"
            title="Login failed"
            size="small"
          >
            {error}
          </Alert>
        )}
        <Formik
          onSubmit={handleLogin}
          validationSchema={LoginSchema}
          initialValues={generateInitialValuesFromSchema(
            Object.keys(LoginSchema.fields)
          )}
          render={props => <LoginForm {...props} loginError={error} />}
        />
      </Card>
    </CardContainer>
  </FullScreenFlex>
);

Login.propTypes = {
  /** Function to handle login */
  handleLogin: PropTypes.func.isRequired,
  loading: PropTypes.bool,
  error: PropTypes.string,
};

Login.defaultProps = {
  loading: false,
  error: null,
};

export default Login;
