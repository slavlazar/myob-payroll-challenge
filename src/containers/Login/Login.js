import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';

import actions from '../../actions/auth';
import { isLoggedInSelector } from '../../selectors/auth';
import Login from '../../components/Login';

const { doLogin } = actions;

/**
 * Login container to handle connection to redux store
 * and handle Login component actions
 * @param {object} Params from Redux store of Auth
 */
export const LoginContainer = ({
  Auth: { loading, isLoggedIn, error },
  dispatchDoLogin,
}) => {
  const handleLogin = ({ username: email, password }) => {
    dispatchDoLogin({ email, password });
  };

  if (isLoggedIn) {
    return <Redirect to="/" />;
  }

  return <Login loading={loading} handleLogin={handleLogin} error={error} />;
};

LoginContainer.propTypes = {
  /** Auth state from redux store */
  Auth: PropTypes.shape({
    loading: PropTypes.bool,
    isLoggedIn: PropTypes.bool,
    error: PropTypes.string,
  }).isRequired,
  /** Function to login from store action creators */
  dispatchDoLogin: PropTypes.func.isRequired,
};

const mapStateToProps = state => {
  return {
    Auth: {
      ...isLoggedInSelector(state),
    },
  };
};

export default connect(
  mapStateToProps,
  {
    dispatchDoLogin: doLogin,
  }
)(LoginContainer);
