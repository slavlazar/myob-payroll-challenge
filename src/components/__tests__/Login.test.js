import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';

import Login from '../Login/Login';
import LoginForm from '../Login/Login.form';

const handleLogin = sinon.spy();

const shallowLoginWrapper = (props = {}) => {
  const allProps = {
    handleLogin,
    ...props,
  };

  return shallow(<Login {...allProps} />);
};

const defaultFormProps = {
  values: {},
  touched: {},
  handleSubmit: handleLogin,
  handleChange: sinon.spy(),
  handleBlur: sinon.spy(),
  isValid: false,
};

const shallowLoginFormWrapper = (props = {}) => {
  const allProps = {
    ...defaultFormProps,
    ...props,
  };

  return shallow(<LoginForm {...allProps} />);
};

describe('Components: Login', () => {
  describe('Login', () => {
    it('renders correctly', () => {
      const wrapped = shallowLoginWrapper();

      expect(wrapped).toMatchSnapshot();
    });

    it('should render an alert for any errors', () => {
      const error = 'Failed to login';
      const wrapped = shallowLoginWrapper({ error });
      const alertBoxWrapped = wrapped.find('.login-alert');
      expect(alertBoxWrapped.dive().text()).toContain(error);
    });
  });

  describe('Login Form', () => {
    afterEach(() => {
      defaultFormProps.handleSubmit.resetHistory();
      defaultFormProps.handleChange.resetHistory();
      defaultFormProps.handleBlur.resetHistory();
    });

    it('renders correctly', () => {
      const wrapped = shallowLoginFormWrapper();

      expect(wrapped).toMatchSnapshot();
    });

    it('should trigger onChange event for input fields', () => {
      const wrapped = shallowLoginFormWrapper();

      const usernameField = wrapped.find('#username');
      const passwordField = wrapped.find('#password');
      usernameField.simulate('change', 'test');
      expect(defaultFormProps.handleChange.callCount).toEqual(1);

      passwordField.simulate('change', 'test2');
      expect(defaultFormProps.handleChange.callCount).toEqual(2);
    });

    it('should trigger the onBlur event for input fields', () => {
      const wrapped = shallowLoginFormWrapper();

      expect(defaultFormProps.handleBlur.callCount).toEqual(0);

      const usernameField = wrapped.find('#username');
      const passwordField = wrapped.find('#password');
      usernameField.simulate('blur');
      expect(defaultFormProps.handleBlur.callCount).toEqual(1);

      passwordField.simulate('blur');
      expect(defaultFormProps.handleBlur.callCount).toEqual(2);
    });

    it('should trigger the onSubmit event', () => {
      const wrapped = shallowLoginFormWrapper({});

      expect(defaultFormProps.handleSubmit.callCount).toEqual(0);

      const button = wrapped.find('#login-form-button');
      button.simulate('click');
      expect(defaultFormProps.handleSubmit.callCount).toEqual(1);
    });

    it('should disable the login button', () => {
      const wrapped = shallowLoginFormWrapper({ isSubmitting: true });

      const button = wrapped.find('#login-form-button');
      expect(button.prop('disabled')).toBe(true);
    });
  });
});
