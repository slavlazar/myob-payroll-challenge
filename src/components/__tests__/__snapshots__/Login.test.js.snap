// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components: Login Login Form renders correctly 1`] = `
<Form>
  <Input
    errorMessage={null}
    id="username"
    label="Username"
    name="username"
    onBlur={[Function]}
    onChange={[Function]}
    type="text"
  />
  <Input
    errorMessage={null}
    id="password"
    label="Password"
    name="password"
    onBlur={[Function]}
    onChange={[Function]}
    type="password"
  />
  <Button
    aria-label="Sign in"
    block={true}
    colour="primary"
    disabled={true}
    id="login-form-button"
    onClick={[Function]}
    size="medium"
    type="submit"
  >
    Sign in
  </Button>
</Form>
`;

exports[`Components: Login Login renders correctly 1`] = `
<styled.main>
  <styled.div>
    <Logo
      alt="MYOB"
      className=""
    />
  </styled.div>
  <Heading
    noMargin={false}
    textAlign="center"
    type="h1"
  >
    Sign in to Pay Slip Generator
  </Heading>
  <styled.div>
    <Card
      loading={false}
    >
      <Formik
        enableReinitialize={false}
        initialValues={
          Object {
            "password": "",
            "username": "",
          }
        }
        isInitialValid={false}
        onSubmit={[Function]}
        render={[Function]}
        validateOnBlur={true}
        validateOnChange={true}
        validationSchema={
          ObjectSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_defaultDefault": [Function],
            "_deps": Array [],
            "_excludedEdges": Array [],
            "_exclusive": Object {},
            "_mutate": undefined,
            "_nodes": Array [
              "password",
              "username",
            ],
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "object",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "fields": Object {
              "password": StringSchema {
                "_blacklist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_conditions": Array [],
                "_deps": Array [],
                "_exclusive": Object {
                  "required": true,
                },
                "_mutate": undefined,
                "_options": Object {
                  "abortEarly": true,
                  "recursive": true,
                },
                "_type": "string",
                "_typeError": [Function],
                "_whitelist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "tests": Array [
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
              },
              "username": StringSchema {
                "_blacklist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_conditions": Array [],
                "_deps": Array [],
                "_exclusive": Object {
                  "required": true,
                },
                "_mutate": undefined,
                "_options": Object {
                  "abortEarly": true,
                  "recursive": true,
                },
                "_type": "string",
                "_typeError": [Function],
                "_whitelist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "tests": Array [
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
              },
            },
            "tests": Array [],
            "transforms": Array [
              [Function],
            ],
          }
        }
      />
    </Card>
  </styled.div>
</styled.main>
`;
